docs/src/proposals/scrambled-tx.md

```diff
@@ -0,0 +1,287 @@
+Thus, their strategy for most likely favorable trade is submission of their
+transaction as fast as possible as determined.
+
+Lastly, transactions are now included into a block very greedily. And
+observation by the cluster can be signalled very reliably. So, clients (esp.
+bots) are less incentivised to spamming for the race of inclusion.
+
+#### Leaders
+
+They want to maximize profit from running the validator in any possible ways.
+So, they try to stall the block production for MEV with the help of searchers.
+However, stalling block production is guaranteed to result in skipped block. All
+in all, this behavior incurs opportunity cost. Thus, their optimal behavior is
+include transactions indiscriminately as many as possible for maximum block
+rewards. This naturally results in the FIFO ordering of transactions.
+
+#### Validators
+
+They want cluster fork less frequently so generally cooperates.
+Unless significant nodes (like more than 1/3) are colluding, they pre-vote
+blocks honestly with accurate wall clock in the fear of reduced rewards because
+of being outstanding outlier otherwise.
+
+Also they're incentivised for publishing scrambled transactions as first solver
+by small rewards. Lying the results is eligible for slashing to ensure bad
+scrambled transaction is detected and backed by stakes.
+
+### Further extensions
+
+This proposal lays down the foundation for further protocol extension.
+
+Firstly, this whole new descrambling process can be regarded as a deterministic
+and async commit-and-reveal. This means a non-interactive, unbiasble,
+deterministic and free random number can be provided at the consensus layer as
+by-product, with a few tweaks of this proposal.
+
+That random can easily be fed into transaction runtime for one thing. Further,
+this random number is also used to combat against verifiers's dilemma (or
+tragedy of commons) as well, because of the very existence at the consensus
+layer, opening fierce yet cooperative validator competition.
+
+All in all, the induced latency should be worth as the trade-off to realize
+these technical breakthroughs.
+
+### Prior work
+
+None? This scrambling is leveraging the fact Solana's very short block time and
+abundance of online GPU resources.
+
+### Remaining concerns
+
+- Unsolved MEV opportunity on the on-chain mempool. So clients should be wise to
+  value their transaction with adequate additional priority fee.
+
+- Reveal of intension from failed scrambled transaction. For that concern,
+  strong (= finger-printing resistant) scrambling is needed. also, this can tip
+  about more-or-less immediate general market sentiment change after some news
+  release, even if these transactions are individually protected from
+  reordering?
+
+- sha256-ing throughput difference between cpu and gpu?
+
+- mev searchers will compete on the on-chain mempool with on-chain transaction
+  submission mechanism as new battle frontier.
```

Comment:
here

Reply:
A leader can still front-run transactions in this scheme. When it receives encryption keys and sees something it would like to front-run it can drop this transaction from the block, pretending it missed that decryption packet. And then send its own transaction to TPU forward. Clients will need to spam to future leaders to combat this (but they do it anyway, haha).

Reply:
well, the decryption (de-scrambling stage) only happens after the scrambled transaction is committed into a block already via adaptive difficulty. so censoring the description packet can't prevent the execution of the original transaction in the block. if the solver node isn't seeing the censoring leader broadcast their description packet, they just retry with future leaders.

> And then send its own transaction to TPU forward.

this is certainly doable, but that copied transaction just executed later and less-favorable.

> Clients will need to spam to future leaders to combat this (but they do it anyway, haha).

after this proposal, clients can immediately be notified of inclusion to a block, pending execution. so, hopefully, they stop doing this as is today. ;)

Reply:
so, what happens then if decryption packet is indeed dropped, entire block is skipped and passed to the next leader? and what if transaction after descrambling is garbage, how it can be committed?

Reply:
@askibin entire block won't be skipped in that case as long as the block is committed by cluster with `pre-vote`. decryption packets are propagated eventually across cluster even under abnormal conditions, unless significant of stake is censoring them.

and the content of block is `pre-vote` committed before decryption _as-is_ and _in scrambled state_. so, blocks might contain garbage indeed. that just will be discard at the description stage. admittedly, that's another side of this proposal's downside. however, dos-ing or spamming can be mounted with/without proposal. so i think this is still tolerable.

also, i know it isn't intuitive compared to other chains, however solana's block processing is all pipelined. so, next leaders will be creating child scrambled blocks for their maximum reward when de-scrambling of the previous block is ongoing.